File: GFG PLACEMENT-100/Greedy/JAVA Solution/Activity_Selection.java
//Activity Selection
//https://practice.geeksforgeeks.org/problems/activity-selection-1587115620/1/


//Method-1 (Greedy Approach)
// { Driver Code Starts
import java.io.*; 
import java.util.*; 
import java.lang.*;

class Main
{
    public static void main (String[] args) throws IOException  
    {
        
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		//testcases
		int t = Integer.parseInt(br.readLine().trim());
		
		while(t-- > 0)
		{
		    String inputLine[] = br.readLine().trim().split(" ");
		    //size of array
		    int n = Integer.parseInt(inputLine[0]);
		    int start[] = new int[n];
		    int end[] = new int[n];
		    
		    //adding elements to arrays start and end
		    inputLine = br.readLine().trim().split(" ");
		    for(int i = 0; i < n; i++)
		     start[i] = Integer.parseInt(inputLine[i]);
		    
		    inputLine = br.readLine().trim().split(" ");
		    for(int i= 0; i < n; i++)
		      end[i] = Integer.parseInt(inputLine[i]);
		    
		    //function call
		    System.out.println(new Solution().activitySelection(start, end, n));
		}
    }
}

// } Driver Code Ends


class Solution
{
    //Function to find the maximum number of activities that can
    //be performed by a single person.
    
    static class Pair implements Comparable<Pair>{
        int start, end;
        Pair(int start, int end){
            this.start=start;
            this.end=end;
        }
        
        public int compareTo(Pair p){
            return this.end-p.end;
        }
    }
    
    
    public static int activitySelection(int start[], int end[], int n)
    {
        // add your code here
        ArrayList<Pair> arr = new ArrayList<>();
        for(int i=0;i<n;i++){
            Pair p= new Pair(start[i],end[i]);
            arr.add(p);
        }
        
        Collections.sort(arr);
        
        int res=1, prev=arr.get(0).end;
        for(int i=1;i<n;i++){
            if(arr.get(i).start > prev){
                res++;
                prev=arr.get(i).end;
            }
        }
        return res;
    }
}

File: GFG PLACEMENT-100/Greedy/CPP Solution/Activity_Selection.cpp
//Activity Selection
//https://practice.geeksforgeeks.org/problems/activity-selection-1587115620/1/


//Method-1 (Greedy Approach)
// { Driver Code Starts
#include<bits/stdc++.h>
using namespace std;


 // } Driver Code Ends
class Solution
{
    public:
    //Function to find the maximum number of activities that can
    //be performed by a single person.
    int activitySelection(vector<int> start, vector<int> end, int n)
    {
        // Your code here
        vector<pair<int,int>> arr;
        for(int i=0;i<n;i++){
            arr.push_back({start[i],end[i]});
        }
        
        sort(arr.begin(),arr.end(),myCmp);
        
        int res=1, prev=0;
        for(int i=1;i<n;i++)
            if(arr[i].first>arr[prev].second){
                res++;
                prev=i;
            }
        
        return res;
    }
    
    static bool myCmp(pair<int,int> a, const pair<int,int> b){
        return a.second<b.second;
    }
};

// { Driver Code Starts.
int main()
{
    int t;
    
    //testcases
    cin >> t;
    while(t--)
    {
        //size of array
        int n;
        cin >> n;
        vector<int> start(n), end(n);
        
        //adding elements to arrays start and end
        for(int i=0;i<n;i++)
            cin>>start[i];
        for(int i=0;i<n;i++)
            cin>>end[i];
        Solution ob;
        //function call
        cout << ob.activitySelection(start, end, n) << endl;
    }
    return 0;
}
  // } Driver Code Ends




//Method-2 (Greedy Approach)

  // { Driver Code Starts
#include<bits/stdc++.h>
using namespace std;


 // } Driver Code Ends
class Solution
{
    public:
    //Function to find the maximum number of activities that can
    //be performed by a single person.
    int activitySelection(vector<int> start, vector<int> end, int n)
    {
        // Your code here
        vector<pair<int,int>> arr;
        for(int i=0;i<n;i++){
            arr.push_back({end[i], start[i]});
        }
        
        //stored end as first and start as second.. so that camparision function is not required
        //storing based on first value of pair (which is end time)
        sort(arr.begin(),arr.end());
        
        int res=1, prev=0;
        for(int i=1;i<n;i++)
            if(arr[i].second>arr[prev].first){
                res++;
                prev=i;
            }
        
        return res;
    }
    
};

// { Driver Code Starts.
int main()
{
    int t;
    
    //testcases
    cin >> t;
    while(t--)
    {
        //size of array
        int n;
        cin >> n;
        vector<int> start(n), end(n);
        
        //adding elements to arrays start and end
        for(int i=0;i<n;i++)
            cin>>start[i];
        for(int i=0;i<n;i++)
            cin>>end[i];
        Solution ob;
        //function call
        cout << ob.activitySelection(start, end, n) << endl;
    }
    return 0;
}
  // } Driver Code Ends